gguf-py: Refactor and allow reading/modifying existing GGUF files

KerfuffleV2 opened the pull request:
This is a big one.

1. Splits `gguf` Python package into logical modules.
2. Makes the constant keys hierarchical (old constants left for compatibility)
3. Some cleanups in the writer, especially when dealing with the endian struct format stuff.
4. I realized the local GGUF thing in the scripts adding `gguf-py/gguf` was silly because that makes it load `gguf.py` specifically, rather than `gguf` as a package. Fixed that.
5. Finally, GGUF file reading support!

The GGUF file reading is done using `numpy`'s `memmap` feature. It also supports endian swapped views. In addition, you can map the file read/write and it's possible to use `GGUFReader` to get a writeable view of the GGUF data (obviously it's necessary to be very careful when writing it, and currently doing anything that changes the size of a field isn't really possible). However, doing something like being able to change a scalar value without rewriting the whole file is nice.

There is an example of dumping a GGUF file at the end of `gguf_reader.py`. Here is example output:

*edit: This is now a separate script in `gguf-py/scripts/gguf-dump.py`*

```plaintext
$ python gguf-py/gguf/gguf_reader.py /blah/yi-q4_k_m.gguf
* Loading: /blah/yi-q4_k_m.gguf

* Dumping 23 key/value pair(s)
      1: UINT32     |        1 | GGUF.version = 3
      2: UINT64     |        1 | GGUF.tensor_count = 543
      3: UINT64     |        1 | GGUF.kv_count = 20
      4: STRING     |        1 | general.architecture = 'llama'
      5: STRING     |        1 | general.name = 'LLaMA v2'
      6: UINT32     |        1 | llama.context_length = 4096
      7: UINT32     |        1 | llama.embedding_length = 7168
      8: UINT32     |        1 | llama.block_count = 60
      9: UINT32     |        1 | llama.feed_forward_length = 20480
     10: UINT32     |        1 | llama.rope.dimension_count = 128
     11: UINT32     |        1 | llama.attention.head_count = 56
     12: UINT32     |        1 | llama.attention.head_count_kv = 8
     13: FLOAT32    |        1 | llama.attention.layer_norm_rms_epsilon = 9.999999747378752e-06
     14: FLOAT32    |        1 | llama.rope.freq_base = 5000000.0
     15: UINT32     |        1 | general.file_type = 15
     16: STRING     |        1 | tokenizer.ggml.model = 'llama'
     17: [STRING]   |    64000 | tokenizer.ggml.tokens
     18: [FLOAT32]  |    64000 | tokenizer.ggml.scores
     19: [INT32]    |    64000 | tokenizer.ggml.token_type
     20: UINT32     |        1 | tokenizer.ggml.bos_token_id = 1
     21: UINT32     |        1 | tokenizer.ggml.eos_token_id = 2
     22: UINT32     |        1 | tokenizer.ggml.padding_token_id = 0
     23: UINT32     |        1 | general.quantization_version = 2

* Dumping 543 tensor(s)
      1:  458752000 |  7168, 64000,     1,     1 | Q4_K    | token_embd.weight
      2:   51380224 |  7168,  7168,     1,     1 | Q4_K    | blk.0.attn_q.weight
      3:    7340032 |  7168,  1024,     1,     1 | Q4_K    | blk.0.attn_k.weight
      4:    7340032 |  7168,  1024,     1,     1 | Q6_K    | blk.0.attn_v.weight
      5:   51380224 |  7168,  7168,     1,     1 | Q4_K    | blk.0.attn_output.weight
      6:  146800640 |  7168, 20480,     1,     1 | Q4_K    | blk.0.ffn_gate.weight
      7:  146800640 | 20480,  7168,     1,     1 | Q6_K    | blk.0.ffn_down.weight
      8:  146800640 |  7168, 20480,     1,     1 | Q4_K    | blk.0.ffn_up.weight
      9:       7168 |  7168,     1,     1,     1 | F32     | blk.0.attn_norm.weight
     10:       7168 |  7168,     1,     1,     1 | F32     | blk.0.ffn_norm.weight

```

*edit*: I wanted to change the BOS token in my Yi model to `2` instead of `1` (their tokenizer config actually says don't add BOS but we don't respect it). Making that change is pretty easy now:

```python
import gguf
reader = gguf.GGUFReader('/path/models/yi-q4_k_m.gguf', 'r+')
reader.fields['tokenizer.ggml.bos_token_id'].parts[-1][0] = 2
```

***

*edit*: I added some examples demonstrating the reader stuff. With this pull checked out, you can run:

    python gguf-py/examples/dump_gguf.py /path/model.gguf

to dump the metadata in a GGUF file.

There is also an example to allow changing simple metadata values in a GGUF file (I.E. integers, floats). Here's an example of changing the BOS token id to `1`:

    python gguf-py/examples/modify_gguf.py /path/model.gguf tokenizer.ggml.bos_token_id 1

Note: It will prompt you if you're sure before actually making any changes. That example only supports simple values, however the `GGUFReader` API will let you change anything (although the GGUF format it self currently prevents changes where the length of a field would be different).

KerfuffleV2 committed:
gguf-py: Refactor and add file reading support

KerfuffleV2 commented:
Killl meee. Of course there was a major change to exactly `gguf.py` in this 2-3 hour window.

KerfuffleV2 commented:
@cebtenzzre I'm not really sure how to resolve these conflicts without basically reverting then manually editing in your changes since I moved the file content and then changed it. Would you be okay with that?

Or is there a better way?

cebtenzzre commented:
> I'm not really sure how to resolve these conflicts without basically reverting then manually editing in your changes since I moved the file content and then changed it. Would you be okay with that?

Yeah, that's basically the only way to do it.

monatis commented on gguf-py/gguf/gguf.py:
I originally wrote the package separated to individual files like this, but then gg wanted to merge them in a single `gguf.py` file. not sure if they like it.

KerfuffleV2 replied:
I guess we'll find out. A fair amount of stuff has been added to it since the earlier versions so maybe making it modular is more compelling now...

If I have to, I can squash everything together again.

KerfuffleV2 committed:
Replay changes from #3871

Credit to @cebtenzzre for that pull

cebtenzzre commented on gguf-py/gguf/gguf.py:
If we're going to install an example, it probably should have a clearer name than `gguf`. And we could move the imports out of `if __name__ == "__main__":`.

cebtenzzre replied:
If you're not intending to run the script in this file, you shouldn't be importing it at all - it has no public API.

KerfuffleV2 replied:
Are you only talking about `gguf.py` here? If so, then sure, I agree. It doesn't actually do anything anymore so it won't matter where the imports are.

KerfuffleV2 replied:
I solved this by just moving the examples into separate files in `examples/`

I also added an example for modifying GGUF files.

What do you think of the current revision?

cebtenzzre commented on gguf-py/gguf/constants.py:
```diff
@@ -0,0 +1,469 @@
+    SOURCE_URL           = "general.source.url"
+    SOURCE_HF_REPO       = "general.source.huggingface.repository"
+    FILE_TYPE            = "general.file_type"
+
```
What are you trying to do here? NamedTuple is like a tuple (i.e., an immutable collection with values you commonly access by index/position)  but with the bonus of having names for its fixed set of fields. I think you want StrEnum.

cebtenzzre commented on gguf-py/gguf/gguf_reader.py:
```diff
@@ -0,0 +1,270 @@
+
+    # Data parts. Some types have multiple components, such as strings
+    # that consist of a length followed by the string data.
+    parts: [npt.NDArray] = []
```
This is spelled `list[npt.NDArray]`. You may want to run mypy to find more errors like this.

KerfuffleV2 replied:
I thought I had it set up properly, but apparently not. These last set of changes clear up almost all the lints even running with `--strict`.

I really appreciate you taking the time to review the pull but please, _please_ don't just add commits to my pulls without talking to me about it first.

I guess I obliterated your changes, although I wasn't really trying to do that. I'm not really sure what the alternative is though since I had a big set of changes that now conflicted with my own pull.

cebtenzzre replied:
Noted. I personally prefer history-preserving operations like merges, it makes history easier to follow. Especially after a PR makes it out of draft status. It'll all get squashed in the end anyway.

I can save my cleanup for whenever you're ready. I'd rather not give you a pile of linter suggestions to correct if you don't have the same linter set up that I do.

KerfuffleV2 replied:
> I can save my cleanup for whenever you're ready. 

You can go ahead (and thanks) at least if you see this in the next few hours. I won't be working on it for a while, gotta sleep.

The main thing I'm concerned about is wanting to push something of my own and running into conflicts. I really don't enjoy dealing with merge conflicts. That's the main thing, but also it's not guaranteed I'll agree with whatever changes so I'd also like to have some idea of what's going to happen in advance.

> I personally prefer history-preserving operations like merges

I agree it's better. I've had some issues with other approaches in the past though, that's why I settled on my (less than idea) workflow of just squashing everything and rebasing. I would definitely prefer to keep the history though.

KerfuffleV2 committed:
Various type annotation fixes.

cebtenzzre committed:
sort imports with isort (again)

cebtenzzre commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -0,0 +1,403 @@
+        if self.temp_file is None:
+            self.tensors.append(tensor)
+
+        tensor.tofile(self.temp_file)
```
This isn't right - temp_file can be None here. It looks like you dropped a `return` statement that was there before.

KerfuffleV2 replied:
Thanks for catching that. You're correct, I did mess it up.

cebtenzzre commented on gguf-py/gguf/constants.py:
```diff
@@ -0,0 +1,469 @@
+    USE_PARALLEL_RESIDUAL: str = "{arch}.use_parallel_residual"
+    TENSOR_DATA_LAYOUT   : str = "{arch}.tensor_data_layout"
+
+class Keys(NamedTuple):
```
This should be Enum, not NamedTuple, and you won't need to specify the types that way. Also, you will then use "Keys.GENERAL", not "KEY.GENERAL".

KerfuffleV2 replied:
I actually tried and wasn't successful. If you want to change it (or any type annotations, reorganize the examples, etc) you have my blessing to go ahead and just add the commits. 

I'm really going to bed now, I swear.

KerfuffleV2 committed:
Fix missing return statement in add_tensor

cebtenzzre committed:
style cleanup with flake8

cebtenzzre committed:
fix NamedTuple and Enum usage

KerfuffleV2 committed:
Fix an issue with state init in GGUFReader

Move examples to an examples/ directory

Clean up examples

Add an example of modifying keys in a GGUF file

Update documentation with info on examples

Try to support people importing gguf/gguf.py directly

KerfuffleV2 committed:
Damagage is not a word.

Galunid commented:
Hey, we are going to have a conflict once #3838 gets merged. You'll need to revert your changes for `convert-*-hf-to-gguf.py` scripts, since they'll be deleted. Instead you need to add that change to `convert-hf-to-gguf.py` to resolve it ;)

Thanks for GGUFReader! It was definitely needed https://github.com/ggerganov/llama.cpp/pull/3838#issuecomment-1783980492

KerfuffleV2 commented:
@Galunid 

> Hey, we are going to have a conflict once https://github.com/ggerganov/llama.cpp/pull/3838 gets merged.

Not a problem. Thanks for the heads up! If you wanted to, you could just make that change in #3838, it isn't specific to the other GGUF changes. Just something weird I noticed when I was looking at those scripts. Whichever way you prefer is fine.

KerfuffleV2 commented:
@chenqiny Hi, not sure if you're interested but this pull is supposed to add the ability to read/modify GGUF files and transparently support working with GGUF files created on a different endian than the machine where the script is running. So you should be able to use this to open and change a little endian GGUF file on your big endian machine and vice versa.

I don't really have a good way to test that though since all my machines are LE and I'm not sure where to even find a BE GGUF model.

cebtenzzre commented on gguf-py/examples/modify_gguf.py:
```diff
@@ -0,0 +1,41 @@
+
+if __name__ == '__main__':
+    if len(sys.argv) < 4:
+        print('modify_gguf: Error: Missing arguments. Syntax: modify_gguf.py <filename> <key> <value>', file = sys.stderr)
```
```suggestion
        print(
            'modify_gguf: Error: Missing arguments. Syntax: modify_gguf.py <filename> <key> <value>',
            file = sys.stderr,
        )
```

chenqiny commented:
> @chenqiny Hi, not sure if you're interested but this pull is supposed to add the ability to read/modify GGUF files and transparently support working with GGUF files created on a different endian than the machine where the script is running. So you should be able to use this to open and change a little endian GGUF file on your big endian machine and vice versa.
> 
> I don't really have a good way to test that though since all my machines are LE and I'm not sure where to even find a BE GGUF model.

Sure. I will test it in weekend.

Galunid commented:
>If you wanted to, you could just make that change in https://github.com/ggerganov/llama.cpp/pull/3838, it isn't specific to the other GGUF changes. Just something weird I noticed when I was looking at those scripts. Whichever way you prefer is fine.

Sure, I included the path fix in #3838

KerfuffleV2 committed:
Clean up gguf-py/examples/modify_gguf.py whitespace

Co-authored-by: Jared Van Bortel <cebtenzzre@gmail.com>

KerfuffleV2 committed:
Update gguf-py/examples/modify_gguf.py formatting

Co-authored-by: Jared Van Bortel <cebtenzzre@gmail.com>

KerfuffleV2 committed:
Update gguf-py/gguf/gguf_reader.py type hint

Co-authored-by: Jared Van Bortel <cebtenzzre@gmail.com>

KerfuffleV2 committed:
Make examples executable, formatting changes

KerfuffleV2 commented on gguf-py/examples/dump_gguf.py:
```diff
@@ -0,0 +1,41 @@
+
+from gguf import GGUFReader, GGUFValueType  # noqa: E402
+
+def dump_gguf(filename: str) -> None:
```
Thanks for all the suggestions (and especially the actual bugs you caught). I really appreciate the time you've spent helping improve this pull!

What are you using for formatting and would you be able to share your configuration? I'd be perfectly happy to turn on Python auto formatting if there's a standard for the Python code in this repo to follow.

cebtenzzre replied:
The linters I'm using are:
- mypy for type checking (there is a mypy.ini in the repo and all new python scripts should pass mypy)
- isort for import sorting (for this repo, basically `isort **/*.py -l 120 --tc -m VERTICAL_HANGING_INDENT`)
- and flake8 for PEP 8 style checking.

My flake8 configuration is messy, but I've done `pip install wemake-python-styleguide` and then turned off everything I don't care about. This ridiculous command should reproduce the way I'm using flake8 for llama.cpp (most of this is hidden behind a shell alias):
```
flake8 **/*.py --max-line-length=120 --ignore=D,DAR,I,S,A003,E121,E123,E126,E127,E201,E202,E203,E211,E221,E222,E226,E241,E251,E261,E266,E272,E306,E402,E704,E731,E741,E800,F403,F811,N400,N801,N803,N806,N812,N813,P101,P103,P205,Q000,T001,U101,W503,W504,WPS102,WPS110,WPS111,WPS113,WPS114,WPS115,WPS117,WPS120,WPS122,WPS125,WPS201,WPS202,WPS203,WPS204,WPS210,WPS211,WPS212,WPS213,WPS214,WPS218,WPS220,WPS221,WPS222,WPS223,WPS224,WPS225,WPS226,WPS229,WPS230,WPS231,WPS232,WPS234,WPS235,WPS236,WPS237,WPS238,WPS300,WPS301,WPS302,WPS304,WPS305,WPS306,WPS316,WPS317,WPS318,WPS319,WPS320,WPS322,WPS323,WPS326,WPS331,WPS332,WPS336,WPS337,WPS347,WPS348,WPS352,WPS360,WPS361,WPS362,WPS400,WPS405,WPS407,WPS412,WPS414,WPS420,WPS421,WPS422,WPS427,WPS428,WPS429,WPS430,WPS431,WPS432,WPS433,WPS434,WPS435,WPS436,WPS437,WPS440,WPS441,WPS442,WPS450,WPS457,WPS458,WPS459,WPS460,WPS463,WPS464,WPS501,WPS504,WPS508,WPS509,WPS510,WPS513,WPS518,WPS526,WPS602,WPS604,WPS605,WPS606,WPS608,WPS609,WPS611,WPS613
```

There is a lot of subjectivity with flake8, even that command leaves some checks enabled that don't really matter IMO. And normally I leave E251 enabled, but the style in this repo seems to use spaces around '=' in keyword arguments.

KerfuffleV2 commented:
@ggerganov Requested a review just to make sure you're basically okay with the direction in this pull (it's getting close to ready) and don't need any major changes. I won't take an approval to mean "merge immediately!"

KerfuffleV2 committed:
Add more information to GGUFReader and examples comments

KerfuffleV2 committed:
Include a gguf Python package version bump

KerfuffleV2 committed:
Merge branch 'master' into feat-gguf-py-read-refactor

KerfuffleV2 committed:
Add convert-gguf-endian.py script

KerfuffleV2 commented:
I added a script for converting the endianness of GGUF files. It supports tensors in `F32`, `F16` and `Q8_0` formats (so anything `convert.py` can produce).

I was able to test that `GGUFReader` does correctly handle non-native endian files. Dumping the converted (now big endian) file on my little endian system worked fine. I also converted it back to little endian and it had the same SHA256 as the original.

cebtenzzre commented:
I'd like to make another cleanup commit (F401, which I forgot that I usually check separately, and convert-gguf-endian.py) if that's alright with you.

KerfuffleV2 commented:
@cebtenzzre 

> I'd like to make another cleanup commit

No problem, thanks for checking!

I think this pull is ready now unless there are further issues/changes required. After the changes you mentioned, is there anything else think needs to be done?

cebtenzzre committed:
cleanup

cebtenzzre commented on gguf-py/pyproject.toml:
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "gguf"
-version = "0.4.6"
+version = "0.4.7"
```
Shouldn't this be a minor version bump since we are adding GGUFReader?

KerfuffleV2 replied:
Sure, do you want to change it or should I? (I assume you mean it should be `0.5,0`?)

cebtenzzre replied:
I changed it.

cebtenzzre committed:
gguf-py : bump minor version

KerfuffleV2 commented:
@cebtenzzre Thanks again for all your help! 

How does merging this tomorrow if no one finds any problems sound?

monatis commented on gguf-py/examples/dump_gguf.py:
```diff
@@ -0,0 +1,45 @@
+
+
+if __name__ == '__main__':
+    if len(sys.argv) < 2:
```
What about using `argparse` to work with CLI args --it gives help message and argument checking for free. Using `sys.args` directly is not a good practice in Python most of the time.


And if we expect these executables to serve users as an actual tool rather than simply an example to demonstrate the usage, we can install them as entry points in `pyproject.toml`, so users can invoke them easily on the command line after pip-installing.

KerfuffleV2 replied:
Thanks for the comment!

Sure, that's a reasonable suggestion and won't make those examples much more complicated. I was trying to hit a balance between usable tool and an example that was still simple enough to be helpful (I actually think I went too far trying to make the output pretty).

Also, for `llama.cpp` people, something hidden away in `gguf-py/examples` isn't really going to be that discoverable. So I'm not 100% sure where those should be. Should the endian converter script also be there?

monatis replied:
Yes, I think argparse can make it even more readable and intuitive for Python people.

For the entry point side, they look to me more like simple but helpful utilities rather than API examples --except for the writing example.

I think those ones except for the writing script can be moved to `scripts` folder and then installed in `pyproject.toml`.

KerfuffleV2 replied:
We probably should use a more consistent naming convention if it's getting installed into the user's path. How about something like this (all under `gguf-py/scripts/`):

1. `gguf-dump.py`
2. `gguf-set-kv.py`
3. `gguf-convert-endian.py`

I don't really have strong feelings about the name so if anyone else wants something different, that's probably not an issue. I do think they should start with a common prefix though.

monatis replied:
Prefixing with `gguf-` looks like a good idea. So they can run in two ways:

1. When not installed to the environment: `python path/to/gguf-py/scripts/gguf-dump` (adding Python because shabang is not universally supported)
2. When installed with pip: `gguf-dump`

For the latter, we need to add them in `[tool.poetry.scripts]` in `pyproject.toml`. My clip.cpp Python binding is an [example](https://github.com/monatis/clip.cpp/blob/main/examples/python_bindings/pyproject.toml) (the very end of the file).

KerfuffleV2 replied:
@monatis I updated the scripts to use `argparse` and reorganized stuff a bit. It didn't see clear how to set up the Poetry thing though.

I want it to be possible to run the scripts like:

    python gguf-py/scripts/gguf-dump.py blah.gguf

And that should work without having to install anything, it should just automatically use the repo-local `gguf` package in that case.

It seems like Poetry wants all the scripts to be in a package that doesn't run standalone?

monatis replied:
We're not limited to a single package --we can add a `__init__.py` file to `scripts` and then include `scripts` in `packages` together with `gguf` in `pyproject.toml`:

```
packages = [
    {include = "gguf"},
    {include = "gguf/py.typed"},
    {include = "scripts"},
]
```

 Then we can list them one by one as in the following:

```
[tool.poetry.scripts]
gguf-dump = 'scripts.gguf_dump'
```

Not tested but such a setup should work.

KerfuffleV2 replied:
@monatis 

After much struggling because it can't really handle files with `-` in the name, this is what I came up with. In `pyproject.toml`:

```toml
[tool.poetry.scripts]
gguf-convert-endian = "scripts:gguf_convert_endian_entrypoint"
gguf-dump = "scripts:gguf_dump_entrypoint"
gguf-set-metadata = "scripts:gguf_set_metadata_entrypoint"
```

In `scripts/__init__.py`:

```python
import importlib

gguf_convert_endian_entrypoint = importlib.import_module("scripts.gguf-convert-endian").main
gguf_dump_entrypoint           = importlib.import_module("scripts.gguf-dump").main
gguf_set_metadata_entrypoint   = importlib.import_module("scripts.gguf-set-metadata").main
```

Before I go ahead and commit that, can you live with that approach? I'm not gonna rename my scripts just to make poetry happy, dammit. :)

monatis replied:
ok to me :)

monatis commented on gguf-py/examples/modify_gguf.py:
```diff
@@ -0,0 +1,75 @@
+        print(f'- Key {repr(key)} already set to requested value {current_value}')
+        sys.exit(0)
+    print('*** Warning *** Warning *** Warning **')
+    print('* Changing fields in a GGUF file can damage it. If you are positive then type YES:')
```
I'm quite suspicious of this workflow. We shouldn't allow users to make their file unusable. Instead of a warning, we can check if the new value has the same byte length as the current value. If not, we simply should not make that change and advice creating the model from scratch if possible. Otherwise, I'm almost certain that we'll get multiple issues complaining about it.

KerfuffleV2 replied:
That script only allows changing simple scalar values, so nothing that can result in a different length. Assuming nothing goes wrong while it's running, it won't violate the GGUF format but changing the value of GGUF fields without knowing what you're doing can definitely make the model unusable.

KerfuffleV2 replied:
Can this be resolved, or do you think there's still an issue?

monatis replied:
Looks good, missed the the usage of `scalar_to_np`. Still the message "unsupported type" can be extended but not a big deal as we don't continue.

KerfuffleV2 replied:
Something like "This tool only supports changing simple values, cannot change field BLAH of type TYPE"?

These tools/examples can, of course, have their functionality expanded in the future. This pull is already pretty big though, so I don't want to throw too much stuff into it.

Doing something like making a general purpose tool to inspect/change fields (including stuff like array fields) wouldn't be that difficult.

monatis replied:
> Something like "This tool only supports changing simple values, cannot change field BLAH of type TYPE"

Looks good --making sure that users know what to expect is enough.

> Maybe we shouldn't make it too easy to mess with GGUF files though

This was the main reason why I started this conversation. You know, this Python project is a community-maintained one, and the very main reason of its existence to write GGUF files from scratch. So making it too easy to mess up with GGUF files may bring more problems than it can solve. I agree that it can be further developed by time but believe that they should happen bit-by-bit due to common requirements and with good testing.

KerfuffleV2 replied:
> I agree that it can be further developed by time but believe that they should happen bit-by-bit due to common requirements and with good testing.

Sure. I'm not really planning to expand those scripts in this pull, and if that ever happens in the future then there would have to be some kind of consensus for that pull to get merged.

I think the current approach is reasonable though: you can only use it to change pretty simple values, it prints out what the original value was (so the user can just change it back) _and_ it warns them they should have a backup.

I believe in taking some steps to prevent people from shooting themselves in the foot, but personally I really, really, _really_ hate it when software basically tells me "*Sorry, you can't be trusted to make this decision: I won't let you do that.*"

KerfuffleV2 committed:
Reorganize scripts

KerfuffleV2 committed:
Make GGUFReader endian detection less arbitrary

KerfuffleV2 committed:
Add JSON dumping support to gguf-dump.py

Which I kind of regret now

KerfuffleV2 committed:
A few for gguf-dump.py cleanups

cebtenzzre commented on gguf-py/scripts/gguf-dump.py:
```diff
@@ -0,0 +1,114 @@
+            if itype == GGUFValueType.STRING:
+                curr["value"] = [str(bytes(field.parts[idx]), encoding="utf-8") for idx in field.data]
+            else:
+                curr["value"] = [pv for idx in field.data for pv in field.parts[idx].tolist()],
```
accidental tuple
```suggestion
                curr["value"] = [pv for idx in field.data for pv in field.parts[idx].tolist()]
```

KerfuffleV2 replied:
Thanks. I'm pretty surprised that didn't cause any type checking issues.

KerfuffleV2 committed:
Murder accidental tuple in gguf-py/scripts/gguf-dump.py

Co-authored-by: Jared Van Bortel <cebtenzzre@gmail.com>

cebtenzzre committed:
cleanup

cebtenzzre committed:
constants : remove unneeded type annotations

cebtenzzre committed:
fix python 3.8 compat

KerfuffleV2 committed:
Set up gguf- scripts in pyproject.toml

KerfuffleV2 committed:
And include scripts/__init__.py, derp

monatis reviewed (approved):
lgtm --thanks for taking this! I think it's good to merge.

KerfuffleV2 commented:
> thanks for taking this! I think it's good to merge.

Thanks for taking the time to review it, and also many thanks to cebtenzzre who found a bunch of bugs and cleaned up a lot of stuff! I'll go ahead and merge it later today unless anyone finds problems in the mean time.

KerfuffleV2 commented:
Actually, I found one more thing I'm going to change (after testing that it doesn't break anything). `convert.py`'s Q8_0 quantization code has:

    quantized_dtype = np.dtype([('d', '<f2'), ('qs', 'i1', (32,))])

Forcing the value to little endian. I'm pretty sure that means using `convert.py` to Q8_0 quantize files is going to be broken on big endian machines. Not sure why I did that, but there's no reason to not let it be the native BE.

KerfuffleV2 committed:
convert.py: We can't currently support Q8_0 on big endian.

KerfuffleV2 committed:
gguf-py: SpecialVocab: Always try available sources for special token ids

gguf-py: SpecialVocab: Try to load merges from merges.txt if not in tokenizer.json

gguf-py: SpecialVocab: Add 'add_bos_token' type bools to GGUF metadata
u

KerfuffleV2 commented:
@cebtenzzre @monatis

I made a few extra changes since you approved so I want to double check. I was looking at the code for big endian and Q8_0 and realized it can't really work unless the byteswapping happens before the quantization part. However, currently that's in `gguf` and occurs too late. Q8_0 would have been broken on big endian systems: I just made it so it doesn't even appear as an option in `convert.py`

There are also some changes to `gguf`s `SpecialVocab` class. I was trying to convert a Yi model and it didn't find the special token ids at all. I realized this was because it would only look at `config.json` for `bos_token_id: 1`, etc if `tokenizer.json` or `tokenizer_config.json` did not exist. However, those may only be defined in `config.json`. So I just changed it to check both places. If it found an id already, it won't overwrite that from `config.json` so this change should only have an effect when those special token ids wouldn't have been found previously. I'd call this a bugfix.

I also made it possible to load merges from `merges.txt`. This only occurs if merges are requested (BPE vocab type) _and_ they weren't found in `tokenizer.json`. So this is only a fallback as well.

The last thing I changed is possibly more controversial. I changed `SpecialVocab` to read the `add_bos_token`, etc booleans in `config.json` and add them to the GGUF file metadata like `tokenizer.ggml.add_bos_token` (a boolean). This is because some models don't want you to add a BOS or whatever token, however we currently just always add BOS for SPM models. Right now, that metadata field is just informational. Actually handling it on the C++ side would have to be a different pull.

I also didn't add it to the official GGUF keys constants. (But I can, if people think it's a good idea.) I think we _should_ support at least including that information in the metadata. Models that don't want something like a BOS token prepended to their prompt can have severely degraded quality when you do it anyway.

cebtenzzre committed:
cleanup

cebtenzzre commented on gguf-py/gguf/vocab.py:
```diff
@@ -47,15 +52,49 @@ def add_to_gguf(self, gw: GGUFWriter, quiet: bool = False) -> None:
+        for typ, add in self.add_special_token.items():
+            if not quiet:
+                print(f'gguf: Setting add special token type {typ} to {add}')
+            gw.add_bool(f'tokenizer.ggml.add_{typ}_token', add)
```
I think the format string should be moved to constants.py, and the keys should added to the GGUF spec.

KerfuffleV2 replied:
I'm definitely more than happy to do that, I just didn't want to unilaterally make changes to the GGUF format.

KerfuffleV2 replied:
It's weird we end up with `add_add_bos_token`, `add_add_eos_token` but it doesn't seem like there's a good alternative.

KerfuffleV2 committed:
Promote add_X_token to GGUF metadata for BOS and EOS

monatis commented:
Thanks again @KerfuffleV2  for the good job in this PR and taking into account every review patiently.

chenqiny commented:
@KerfuffleV2 I got following error.

chenqiny@datalake:/database/llama.cpp/gguf-py$ /home/chenqiny/.local/bin/gguf-convert-endian /database/models--meta-llama--Llama-2-7b/snapshots/365ffa8f1a6c455d3e2028ae658236b4b85ba824/ggml-model-f16-little.gguf big
* Loading: /database/models--meta-llama--Llama-2-7b/snapshots/365ffa8f1a6c455d3e2028ae658236b4b85ba824/ggml-model-f16-little.gguf
* Host is LITTLE endian, GGUF file seems to be LITTLE endian
Traceback (most recent call last):
  File "/home/chenqiny/.local/bin/gguf-convert-endian", line 8, in <module>
    sys.exit(gguf_convert_endian_entrypoint())
  File "/home/chenqiny/.local/lib/python3.8/site-packages/scripts/gguf-convert-endian.py", line 109, in main
    convert_byteorder(reader, args)
  File "/home/chenqiny/.local/lib/python3.8/site-packages/scripts/gguf-convert-endian.py", line 34, in convert_byteorder
    if file_endian == order:
UnboundLocalError: local variable 'order' referenced before assignment
